### PR TITLE
Fixer for CA1826

### PR DESCRIPTION
**Pull request by TKharaishvili**
Addresses the issue - https://github.com/dotnet/roslyn-analyzers/issues/1932

The changes in this PR are obviously not ideal, some of the issues being:

1. lack of tests
2. not being implemented for VB
3. the fixer window showing up for other linq methods too even though it's only implemented for `Enumerable.First`

but hey it does work for C# and hopefully you guys can help me deal with the rough edges.

### CODE REVIEW

File: src/Microsoft.NetCore.Analyzers/Core/Runtime/DoNotUseEnumerableMethodsOnIndexableCollectionsInsteadUseTheCollectionDirectly.Fixer.cs [hunk 1]
```diff
@@ -21,9 +29,51 @@ public sealed override FixAllProvider GetFixAllProvider()
 
         public sealed override Task RegisterCodeFixesAsync(CodeFixContext context)
         {
+            var diagnostic = context.Diagnostics[0];
+            var codeAction = CodeAction.Create("Use indexer", c => UseCollectionDirectly(context.Document, context.Span, c));
```
Comment (mavasani) [inline src/Microsoft.NetCore.Analyzers/Core/Runtime/DoNotUseEnumerableMethodsOnIndexableCollectionsInsteadUseTheCollectionDirectly.Fixer.cs]: > "Use indexer"

This needs to be a localized string.

Comment (mavasani) [inline src/Microsoft.NetCore.Analyzers/Core/Runtime/DoNotUseEnumerableMethodsOnIndexableCollectionsInsteadUseTheCollectionDirectly.Fixer.cs]: > CodeAction.Create

Please use the pattern used in this repo for creating code actions, see https://github.com/dotnet/roslyn-analyzers/blob/master/src/Microsoft.CodeQuality.Analyzers/Core/ApiDesignGuidelines/AbstractTypesShouldNotHaveConstructors.Fixer.cs#L59

File: src/Microsoft.NetCore.Analyzers/Core/Runtime/DoNotUseEnumerableMethodsOnIndexableCollectionsInsteadUseTheCollectionDirectly.Fixer.cs [hunk 2]
```diff
@@ -21,9 +29,51 @@ public sealed override FixAllProvider GetFixAllProvider()
...
             return Task.CompletedTask;
 
         }
+
+        private async Task<Document> UseCollectionDirectly(Document document, TextSpan span, CancellationToken c)
+        {
+            var syntaxRoot = await document.GetSyntaxRootAsync(c).ConfigureAwait(false);
+            var semanticModel = await document.GetSemanticModelAsync(c).ConfigureAwait(false);
+
+            var invocationSyntax = GetSyntaxOfType<InvocationExpressionSyntax>(syntaxRoot.FindNode(span));
```
Comment (mavasani) [inline src/Microsoft.NetCore.Analyzers/Core/Runtime/DoNotUseEnumerableMethodsOnIndexableCollectionsInsteadUseTheCollectionDirectly.Fixer.cs]: `syntaxRoot.FindNode(span)` might not succeed and return null, please gracefully bail out for such cases.

Comment (mavasani) [reply src/Microsoft.NetCore.Analyzers/Core/Runtime/DoNotUseEnumerableMethodsOnIndexableCollectionsInsteadUseTheCollectionDirectly.Fixer.cs]: GetSyntaxOfType might not succeed and return null.

File: src/Microsoft.NetCore.Analyzers/Core/Runtime/DoNotUseEnumerableMethodsOnIndexableCollectionsInsteadUseTheCollectionDirectly.Fixer.cs [hunk 3]
```diff
@@ -21,9 +29,51 @@ public sealed override FixAllProvider GetFixAllProvider()
...
+            var invocationOp = (IInvocationOperation)semanticModel.GetOperation(invocationSyntax);
```
Comment (mavasani) [inline src/Microsoft.NetCore.Analyzers/Core/Runtime/DoNotUseEnumerableMethodsOnIndexableCollectionsInsteadUseTheCollectionDirectly.Fixer.cs]: This cast may not always succeed. Please use `as IInvocationOperation` instead.

File: src/Microsoft.NetCore.Analyzers/Core/Runtime/DoNotUseEnumerableMethodsOnIndexableCollectionsInsteadUseTheCollectionDirectly.Fixer.cs [hunk 4]
```diff
@@ -21,9 +29,51 @@ public sealed override FixAllProvider GetFixAllProvider()
...
+
+            //implemented for "Enumerable.First" only
+            if (invocationOp.TargetMethod.Name != "First")
+            {
+                return document;
+            }
+
+            var collectionSyntax = GetSyntaxOfType<ExpressionSyntax>(invocationOp.Arguments[0].Syntax);
```
Comment (mavasani) [inline src/Microsoft.NetCore.Analyzers/Core/Runtime/DoNotUseEnumerableMethodsOnIndexableCollectionsInsteadUseTheCollectionDirectly.Fixer.cs]: Method name being First does not guarantee that it has non-zero arguments. This code needs stronger guard.

Comment (TKharaishvili) [reply src/Microsoft.NetCore.Analyzers/Core/Runtime/DoNotUseEnumerableMethodsOnIndexableCollectionsInsteadUseTheCollectionDirectly.Fixer.cs]: @mavasani this particular check is already done in the analyzer itself:

https://github.com/dotnet/roslyn-analyzers/blob/a44670f717ad86b71f2aeb2aa910c4d726732874/src/Microsoft.NetCore.Analyzers/Core/Runtime/DoNotUseEnumerableMethodsOnIndexableCollectionsInsteadUseTheCollectionDirectly.cs#L138

Does it still need to be checked in the fixer too?

P.S. Thanks for all these reviews. I'll fix them soon.

Comment (mavasani) [reply src/Microsoft.NetCore.Analyzers/Core/Runtime/DoNotUseEnumerableMethodsOnIndexableCollectionsInsteadUseTheCollectionDirectly.Fixer.cs]: Yes, fixers should not assume anything about the analyzer and/or the diagnostic location or properties. For example, you could have a completely different analyzer package reporting the same diagnostic ID, and these diagnostics will also get handed to the fixer.

**Comment by TKharaishvili** [comment]
@mavasani just pushed a new commit with requested fixes as well as unit tests.

The bottom two unit tests named `CA1826FixEnumerableFirstMethodChainCallWrongFormattingCSharp` and `CA1826FixEnumerableFirsProblematicSyntaxCallCSharp` document problematic cases I have a little trouble figuring out and will probably need some help with.

As for:
>Yes, fixers should not assume anything about the analyzer and/or the diagnostic location or properties.

I added a properties dictionary to the analyzer part. This serves two purposes:
1. It makes sure that the fixer window only shows up for `Enumerable.First` and not for the other methods.
2. The fixer now has more guarantees as to which analyzer invoked it, therefore doesn't require as much defensive code.

If you still think that the fixer should contain those guards regardless, let me know and I'll add them in.

Thanks for your help once again.

File: src/Microsoft.NetCore.Analyzers/Core/Runtime/DoNotUseEnumerableMethodsOnIndexableCollectionsInsteadUseTheCollectionDirectly.Fixer.cs [hunk 5]
```diff
@@ -21,9 +29,51 @@ public sealed override FixAllProvider GetFixAllProvider()
 
         public sealed override Task RegisterCodeFixesAsync(CodeFixContext context)
         {
+            var diagnostic = context.Diagnostics[0];
+            var codeAction = CodeAction.Create("Use indexer", c => UseCollectionDirectly(context.Document, context.Span, c));
+            context.RegisterCodeFix(codeAction, diagnostic);
+
             // Fixer not yet implemented.
```
[the remark this answers is not in the document]
Comment (TKharaishvili) [reply src/Microsoft.NetCore.Analyzers/Core/Runtime/DoNotUseEnumerableMethodsOnIndexableCollectionsInsteadUseTheCollectionDirectly.Fixer.cs]: Done in the new commit.

File: src/Microsoft.NetCore.Analyzers/Core/Runtime/DoNotUseEnumerableMethodsOnIndexableCollectionsInsteadUseTheCollectionDirectly.Fixer.cs (hunk 1 above)
Comment (TKharaishvili) [reply src/Microsoft.NetCore.Analyzers/Core/Runtime/DoNotUseEnumerableMethodsOnIndexableCollectionsInsteadUseTheCollectionDirectly.Fixer.cs]: Done in the new commit.

Comment (TKharaishvili) [reply src/Microsoft.NetCore.Analyzers/Core/Runtime/DoNotUseEnumerableMethodsOnIndexableCollectionsInsteadUseTheCollectionDirectly.Fixer.cs]: Done in the new commit.

File: src/Microsoft.NetCore.Analyzers/Core/Runtime/DoNotUseEnumerableMethodsOnIndexableCollectionsInsteadUseTheCollectionDirectly.Fixer.cs (hunk 2 above)
Comment (TKharaishvili) [reply src/Microsoft.NetCore.Analyzers/Core/Runtime/DoNotUseEnumerableMethodsOnIndexableCollectionsInsteadUseTheCollectionDirectly.Fixer.cs]: Done in the new commit.

File: src/Microsoft.NetCore.Analyzers/Core/Runtime/DoNotUseEnumerableMethodsOnIndexableCollectionsInsteadUseTheCollectionDirectly.Fixer.cs (hunk 3 above)
Comment (TKharaishvili) [reply src/Microsoft.NetCore.Analyzers/Core/Runtime/DoNotUseEnumerableMethodsOnIndexableCollectionsInsteadUseTheCollectionDirectly.Fixer.cs]: Done in the new commit.

File: src/Microsoft.NetCore.Analyzers/Core/Runtime/DoNotUseEnumerableMethodsOnIndexableCollectionsInsteadUseTheCollectionDirectly.Fixer.cs (hunk 4 above)
Comment (TKharaishvili) [reply src/Microsoft.NetCore.Analyzers/Core/Runtime/DoNotUseEnumerableMethodsOnIndexableCollectionsInsteadUseTheCollectionDirectly.Fixer.cs]: Addressed in the comment below.

File: src/Microsoft.NetCore.Analyzers/Core/Runtime/DoNotUseEnumerableMethodsOnIndexableCollectionsInsteadUseTheCollectionDirectly.cs [hunk 6]
```diff
@@ -84,7 +84,8 @@ private static void OnCompilationStart(CompilationStartAnalysisContext context)
                     return;
                 }
 
-                operationContext.ReportDiagnostic(Diagnostic.Create(Rule, invocation.Syntax.GetLocation()));
+                var properties = new Dictionary<string, string> { ["method"] = invocation.TargetMethod.Name }.ToImmutableDictionary();
```
Comment (mavasani) [inline src/Microsoft.NetCore.Analyzers/Core/Runtime/DoNotUseEnumerableMethodsOnIndexableCollectionsInsteadUseTheCollectionDirectly.cs]: Can you define a constant string field to hold the key and use the field both here and in the code fixer?

File: src/Microsoft.NetCore.Analyzers/Core/Runtime/xlf/SystemRuntimeAnalyzersResources.cs.xlf [hunk 7]
```diff
@@ -2,6 +2,11 @@
 <xliff xmlns="urn:oasis:names:tc:xliff:document:1.2" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" version="1.2" xsi:schemaLocation="urn:oasis:names:tc:xliff:document:1.2 xliff-core-1.2-transitional.xsd">
   <file datatype="xml" source-language="en" target-language="cs" original="../SystemRuntimeAnalyzersResources.resx">
     <body>
+      <trans-unit id="UseIndexerInstead">
+        <source>Use indexer instead</source>
```
Comment (mavasani) [inline src/Microsoft.NetCore.Analyzers/Core/Runtime/xlf/SystemRuntimeAnalyzersResources.cs.xlf]: Remove “instead”

File: src/Microsoft.NetCore.Analyzers/Core/Runtime/DoNotUseEnumerableMethodsOnIndexableCollectionsInsteadUseTheCollectionDirectly.Fixer.cs [hunk 8]
```diff
@@ -21,9 +30,77 @@ public sealed override FixAllProvider GetFixAllProvider()
 
         public sealed override Task RegisterCodeFixesAsync(CodeFixContext context)
         {
-            // Fixer not yet implemented.
+            // Only one diagnostic expected 
+            Diagnostic diagnostic = context.Diagnostics.Single();
```
Comment (mavasani) [inline src/Microsoft.NetCore.Analyzers/Core/Runtime/DoNotUseEnumerableMethodsOnIndexableCollectionsInsteadUseTheCollectionDirectly.Fixer.cs]: Switch to FirstOrDefault followed by a null check.

File: src/Microsoft.NetCore.Analyzers/Core/Runtime/DoNotUseEnumerableMethodsOnIndexableCollectionsInsteadUseTheCollectionDirectly.Fixer.cs [hunk 9]
```diff
@@ -21,9 +30,77 @@ public sealed override FixAllProvider GetFixAllProvider()
 
         public sealed override Task RegisterCodeFixesAsync(CodeFixContext context)
         {
-            // Fixer not yet implemented.
+            // Only one diagnostic expected
```
Comment (mavasani) [inline src/Microsoft.NetCore.Analyzers/Core/Runtime/DoNotUseEnumerableMethodsOnIndexableCollectionsInsteadUseTheCollectionDirectly.Fixer.cs]: This cannot be assumed given any analyzer can report diagnostic with this ID.

File: src/Microsoft.NetCore.Analyzers/Core/Runtime/DoNotUseEnumerableMethodsOnIndexableCollectionsInsteadUseTheCollectionDirectly.Fixer.cs (hunk 4 above)
Comment (mavasani) [reply src/Microsoft.NetCore.Analyzers/Core/Runtime/DoNotUseEnumerableMethodsOnIndexableCollectionsInsteadUseTheCollectionDirectly.Fixer.cs]: Add a check that you have at least one argument.

File: src/Microsoft.NetCore.Analyzers/UnitTests/Runtime/DoNotUseEnumerableMethodsOnIndexableCollectionsInsteadUseTheCollectionDirectlyTests.Fixer.cs [hunk 10]
```diff
@@ -29,5 +30,186 @@ protected override CodeFixProvider GetCSharpCodeFixProvider()
...
+        var f3 = matrix[0][0];
+        Console.WriteLine(list[0]);
+    }
+
+    IReadOnlyList<int> GetList()
+    {
+        return new List<int> { 1, 2, 3 };
+    }
+}
+", allowNewCompilerDiagnostics: true);
```
Comment (mavasani) [inline src/Microsoft.NetCore.Analyzers/UnitTests/Runtime/DoNotUseEnumerableMethodsOnIndexableCollectionsInsteadUseTheCollectionDirectlyTests.Fixer.cs]: What new diagnostic is generated? Can it be avoided?

File: src/Microsoft.NetCore.Analyzers/UnitTests/Runtime/DoNotUseEnumerableMethodsOnIndexableCollectionsInsteadUseTheCollectionDirectlyTests.Fixer.cs [hunk 11]
```diff
@@ -29,5 +30,186 @@ protected override CodeFixProvider GetCSharpCodeFixProvider()
...
```
Comment (mavasani) [inline src/Microsoft.NetCore.Analyzers/UnitTests/Runtime/DoNotUseEnumerableMethodsOnIndexableCollectionsInsteadUseTheCollectionDirectlyTests.Fixer.cs]: This should be false everywhere. We don’t prefer code fixes to break existing code.

File: src/Microsoft.NetCore.Analyzers/Core/Runtime/DoNotUseEnumerableMethodsOnIndexableCollectionsInsteadUseTheCollectionDirectly.Fixer.cs [hunk 12]
```diff
@@ -21,9 +30,77 @@ public sealed override FixAllProvider GetFixAllProvider()
...
+
+            // The fixer is only implemented for "Enumerable.First"
+            if (!diagnostic.Properties.TryGetValue("method", out var method) || method != "First")
+            {
+                return Task.CompletedTask;
+            }
+
+            string title = SystemRuntimeAnalyzersResources.UseIndexerInstead;
+
+            context.RegisterCodeFix(new MyCodeAction(title,
```
Comment (mavasani) [inline src/Microsoft.NetCore.Analyzers/Core/Runtime/DoNotUseEnumerableMethodsOnIndexableCollectionsInsteadUseTheCollectionDirectly.Fixer.cs]: I think we should offer a code fix only for cases where the compile time type of the invocation receiver/instance has an indexer property with single integer parameter. That way we cover majority of cases and guarantee that the code fix will not break user code.

File: src/Microsoft.NetCore.Analyzers/UnitTests/Runtime/DoNotUseEnumerableMethodsOnIndexableCollectionsInsteadUseTheCollectionDirectlyTests.Fixer.cs (hunk 10 above)
Comment (TKharaishvili) [reply src/Microsoft.NetCore.Analyzers/UnitTests/Runtime/DoNotUseEnumerableMethodsOnIndexableCollectionsInsteadUseTheCollectionDirectlyTests.Fixer.cs]: @mavasani `CS8019: Unnecessary using directive.` It's complaining about `System.Linq` since there's nothing else using it. I think it's pretty harmless.

File: src/Microsoft.NetCore.Analyzers/Core/Runtime/DoNotUseEnumerableMethodsOnIndexableCollectionsInsteadUseTheCollectionDirectly.Fixer.cs (hunk 12 above)
Comment (mavasani) [reply src/Microsoft.NetCore.Analyzers/Core/Runtime/DoNotUseEnumerableMethodsOnIndexableCollectionsInsteadUseTheCollectionDirectly.Fixer.cs]: Can you add pragma disable for it in source instead of blanket disable all diagnostics in new code?

File: src/Microsoft.NetCore.Analyzers/UnitTests/Runtime/DoNotUseEnumerableMethodsOnIndexableCollectionsInsteadUseTheCollectionDirectlyTests.Fixer.cs (hunk 10 above)
Comment (mavasani) [reply src/Microsoft.NetCore.Analyzers/UnitTests/Runtime/DoNotUseEnumerableMethodsOnIndexableCollectionsInsteadUseTheCollectionDirectlyTests.Fixer.cs]: Can you add pragma disable for it in source instead of blanket disable all diagnostics in new code?

Comment (TKharaishvili) [reply src/Microsoft.NetCore.Analyzers/UnitTests/Runtime/DoNotUseEnumerableMethodsOnIndexableCollectionsInsteadUseTheCollectionDirectlyTests.Fixer.cs]: Will do.

File: src/Microsoft.NetCore.Analyzers/Core/Runtime/DoNotUseEnumerableMethodsOnIndexableCollectionsInsteadUseTheCollectionDirectly.Fixer.cs (hunk 12 above)
Comment (TKharaishvili) [reply src/Microsoft.NetCore.Analyzers/Core/Runtime/DoNotUseEnumerableMethodsOnIndexableCollectionsInsteadUseTheCollectionDirectly.Fixer.cs]: As I said above, having the properties check ensures that the fixer was invoked by the correct analyzer. Namely the one that has done all the right checks, making sure that the type of the collection is correct as well as the fact that the `First` function has a single argument.

I know I've asked about this before and you said that it was necessary to duplicate the checks in the fixer. The difference now is that properties check.

So is that sufficient? Or do you still think the checks should be duplicated? If yes, I'll do it right away.

Comment (mavasani) [reply src/Microsoft.NetCore.Analyzers/Core/Runtime/DoNotUseEnumerableMethodsOnIndexableCollectionsInsteadUseTheCollectionDirectly.Fixer.cs]: I got confused into assuming that fix is offered for cases where the type did not have indexers, and hence led to compile time errors. I will resolve this conversation, as that does not seem to be the case.

Comment (mavasani) [reply src/Microsoft.NetCore.Analyzers/Core/Runtime/DoNotUseEnumerableMethodsOnIndexableCollectionsInsteadUseTheCollectionDirectly.Fixer.cs]: I got confused into assuming that fix is offered for cases where the type did not have indexers, and hence led to compile time errors. I will resolve this conversation, as that does not seem to be the case.

**Comment by TKharaishvili** [comment]
@mavasani just pushed the changes addressing the 2nd code review. I'll resolve those conversations now.

File: src/Microsoft.NetCore.Analyzers/Core/Runtime/DoNotUseEnumerableMethodsOnIndexableCollectionsInsteadUseTheCollectionDirectly.Fixer.cs [hunk 13]
```diff
@@ -21,9 +30,81 @@ public sealed override FixAllProvider GetFixAllProvider()
...
+        }
+
+        private T GetSyntaxOfType<T>(SyntaxNode node)
+        {
+            if (!(node is T result))
+            {
+                result = node
+                    .ChildNodes()
+                    .OfType<T>()
+                    .First();
```
Comment (mavasani) [inline src/Microsoft.NetCore.Analyzers/Core/Runtime/DoNotUseEnumerableMethodsOnIndexableCollectionsInsteadUseTheCollectionDirectly.Fixer.cs]:
```suggestion
                    .FirstOrDefault();
```

File: src/Microsoft.NetCore.Analyzers/UnitTests/Runtime/DoNotUseEnumerableMethodsOnIndexableCollectionsInsteadUseTheCollectionDirectlyTests.Fixer.cs [hunk 14]
```diff
@@ -29,5 +30,198 @@ protected override CodeFixProvider GetCSharpCodeFixProvider()
...
+    {
+        GetList()[0];
+    }
+
+    IReadOnlyList<int> GetList()
+    {
+        return new List<int> { 1, 2, 3 };
+    }
+}
+", allowNewCompilerDiagnostics: true, validationMode: TestValidationMode.AllowCompileErrors);
```
Comment (mavasani) [inline src/Microsoft.NetCore.Analyzers/UnitTests/Runtime/DoNotUseEnumerableMethodsOnIndexableCollectionsInsteadUseTheCollectionDirectlyTests.Fixer.cs]: I am fine with this edge case breaking user code, as the original code seems to have a redundant `.First()` invocation. Can you please add a comment about the compiler error generated in the fixed code?

File: src/Microsoft.NetCore.Analyzers/UnitTests/Runtime/DoNotUseEnumerableMethodsOnIndexableCollectionsInsteadUseTheCollectionDirectlyTests.Fixer.cs [hunk 15]
```diff
@@ -29,5 +30,198 @@ protected override CodeFixProvider GetCSharpCodeFixProvider()
         {
             return new CSharpDoNotUseEnumerableMethodsOnIndexableCollectionsInsteadUseTheCollectionDirectlyFixer();
         }
+
+        [Fact]
+        public void CA1826FixEnumerableFirstExtensionCallCSharp()
```
Comment (mavasani) [inline src/Microsoft.NetCore.Analyzers/UnitTests/Runtime/DoNotUseEnumerableMethodsOnIndexableCollectionsInsteadUseTheCollectionDirectlyTests.Fixer.cs]: Can you please add a corresponding VisualBasic test? I think a single VisualBasic test should suffice given the fix code is completely in language agnostic layer.

Comment (TKharaishvili) [reply src/Microsoft.NetCore.Analyzers/UnitTests/Runtime/DoNotUseEnumerableMethodsOnIndexableCollectionsInsteadUseTheCollectionDirectlyTests.Fixer.cs]: The fixer references the `InvocationExpressionSyntax` and `ExpressionSyntax` classes which come from `Microsoft.CodeAnalysis.CSharp.Syntax` namespace. I assume that makes it C# specific.

Do you want me to move those parts to the C# specific subclass? I can attempt to make it work for VB as well?

Comment (mavasani) [reply src/Microsoft.NetCore.Analyzers/UnitTests/Runtime/DoNotUseEnumerableMethodsOnIndexableCollectionsInsteadUseTheCollectionDirectlyTests.Fixer.cs]: Oh I just completely overlooked that - C# specific references should not be accessed in the common layer. That is a layering violation. I am fine with you taking either approach:
1. Move it to C# layer and make the current implementation C#-only, with possible future extension to VB.
2. Add VB/C# specific implementations in respective layers and add unit tests for VisualBasic.

**Comment by TKharaishvili** [comment]
@mavasani I tried to make it work for VB too but wasn't able to. I also opened up a review, please have a look at that as well. Thanks.

File: src/Microsoft.NetCore.Analyzers/UnitTests/Runtime/DoNotUseEnumerableMethodsOnIndexableCollectionsInsteadUseTheCollectionDirectlyTests.Fixer.cs [hunk 16]
```diff
@@ -29,5 +30,198 @@ protected override CodeFixProvider GetCSharpCodeFixProvider()
...
+    }
+}
+");
+        }
+
+        [Fact]
+        public void CA1826FixEnumerableFirstMethodChainCallWrongFormattingCSharp()
+        {
+            //this unit test documents a problematic edge case which needs to be discussed and addressed
+
```
[the remark this answers is not in the document]
Comment (mavasani) [reply src/Microsoft.NetCore.Analyzers/UnitTests/Runtime/DoNotUseEnumerableMethodsOnIndexableCollectionsInsteadUseTheCollectionDirectlyTests.Fixer.cs]: I am fine with leaving this case as it is related to trivia handling and tackle it later as a bug fix based on customer feedback.

[the remark this answers is not in the document]
Comment (TKharaishvili) [reply src/Microsoft.NetCore.Analyzers/UnitTests/Runtime/DoNotUseEnumerableMethodsOnIndexableCollectionsInsteadUseTheCollectionDirectlyTests.Fixer.cs]: ok, should I leave the unit test?

File: src/Microsoft.NetCore.Analyzers/Core/Runtime/DoNotUseEnumerableMethodsOnIndexableCollectionsInsteadUseTheCollectionDirectly.Fixer.cs [hunk 17]
```diff
@@ -19,11 +19,9 @@ public sealed override FixAllProvider GetFixAllProvider()
             return WellKnownFixAllProviders.BatchFixer;
         }
 
-        public sealed override Task RegisterCodeFixesAsync(CodeFixContext context)
```
Comment (mavasani) [inline src/Microsoft.NetCore.Analyzers/Core/Runtime/DoNotUseEnumerableMethodsOnIndexableCollectionsInsteadUseTheCollectionDirectly.Fixer.cs]: I would just delete this file completely. We can extract out a common type type if/when we add VB support.

File: src/Microsoft.NetCore.Analyzers/UnitTests/Runtime/DoNotUseEnumerableMethodsOnIndexableCollectionsInsteadUseTheCollectionDirectlyTests.Fixer.cs (hunk 16 above)
[the remark this answers is not in the document]
Comment (mavasani) [reply src/Microsoft.NetCore.Analyzers/UnitTests/Runtime/DoNotUseEnumerableMethodsOnIndexableCollectionsInsteadUseTheCollectionDirectlyTests.Fixer.cs]: Yes, please keep it, doesn't hurt for it to stay. Thanks!

File: src/Microsoft.NetCore.Analyzers/Core/Runtime/DoNotUseEnumerableMethodsOnIndexableCollectionsInsteadUseTheCollectionDirectly.Fixer.cs (hunk 17 above)
Comment (TKharaishvili) [reply src/Microsoft.NetCore.Analyzers/Core/Runtime/DoNotUseEnumerableMethodsOnIndexableCollectionsInsteadUseTheCollectionDirectly.Fixer.cs]: Done

**Comment by TKharaishvili** [comment]
@mavasani thanks to you too! Couldn't have done this without your guidance.

**Comment by mavasani** [comment]
@TKharaishvili Can you resolve the merge conflicts? I'd start by just resolving the conflict in `src/Microsoft.NetCore.Analyzers/Core/Runtime/SystemRuntimeAnalyzersResources.resx` and then doing a rebuild to regenerate the xlf file.

**Comment by TKharaishvili** [comment]
@mavasani I resolved that file in the github ui. But I'm not sure how to trigger rebuild from here...

**Comment by mavasani** [comment]
> I resolved that file in the github ui. But I'm not sure how to trigger rebuild from here...

Not sure if that is supported. You'll probably have to do it locally by deleting the xlf folder on your machine and rebuilding and pushing it.

**Comment by TKharaishvili** [comment]
@mavasani  I think it's all finished
